code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/dao/impl/ServiceExposeMapDaoImpl.java

```diff
@@ -154,4 +103,52 @@ public void updateEnvironmentName(Environment env) {
                                         CommonStatesConstants.REMOVING))))
                 .fetchInto(ServiceExposeMapRecord.class);
     }
+
+    @Override
+    public List<Service> collectSidekickServices(Service initialService, Map<String, String> initialSvcLabels) {
+        List<Service> servicesToCollect = new ArrayList<>();
+        if (initialSvcLabels == null) {
+            initialSvcLabels = sdService.getServiceLabels(initialService);
+        }
+        String initialSvcSidekickLabel = initialSvcLabels.
+                get(ServiceDiscoveryConstants.LABEL_SERVICE_SIDEKICK) == null ? null : initialSvcLabels
+                .get(ServiceDiscoveryConstants.LABEL_SERVICE_SIDEKICK);
+
+        // add itself
+        servicesToCollect.add(initialService);
+
+        // add sidekick services
+        if (initialSvcSidekickLabel != null) {
+            /*
+             * Find all sidekick services. Just look at sidekick, not volumes-from or other things
+             */
+            List<? extends Service> services = sdService.listEnvironmentServices(initialService.getEnvironmentId());
+            for (Service svc : services) {
+                if (svc.getId().equals(initialService.getId())) {
+                    continue;
+                }
+
+                if (svc.getRemoved() == null) {
```

Comment:
Anytime we have `getRemoved() == null` check you have to consider will the code just get slower and slower over time as the user adds/deletes things.  Sometimes the nature of the code that there will only be a couple, but in this situation it will just keep growing.